### PR TITLE
[WiP][PoC] Lighting

### PR DESCRIPTION
**Pull request by OzoneH3**
Proof of concept for: #23996

Added a "game" to the gamewatch "Experimental Lighting" for testing.

* Extra features: Obstacle transparency and light source strength.

### CODE REVIEW

File: src/lighting.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    while (j < tmpNewIntN) {
+        n = merge1(tmpNew[j], n);
+        j++;
+    }
+    return n;
+}
+
+int Lighting::merge1(T t, int n)
```
Comment (Aivean) [inline src/lighting.cpp]: @OzoneH3 `t` here is modified, so it should be passed by reference.

File: src/lighting.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        }
+    }
+}
+
+void Lighting::applyLight(T t, int x, int y, double td)
+{
+    light[x][y] += max(min(t.b + t.l, x + 1) - max(t.b, x), 0) * t.i * td;
+}
+
+void Lighting::cutInterval(T t, int x)
```
Comment (Aivean) [inline src/lighting.cpp]: @OzoneH3 `t` here is modified, so it should be passed by reference.

**Comment by Aivean** [comment]
@OzoneH3, also, if that is not too much to ask, I'd appreciate if you benchmark C++ port on 100x100 field, completely filled with lightsources. 

I'm curious how it compares to Java implementation speed-wise.

**Comment by OzoneH3** [comment]
@Aivean References did the trick on all occurances of T.
Sure, once it's fully working tests are easy enough to implement
and various unit tests will be required as kevin wrote.

Is there a better way to calculate all directions? I really don't like
that rotate approach thats in the Javascript version.

**Comment by Aivean** [comment]
@OzoneH3, I can think of three approaches for directions:

1. Rotate the input matrix (similar to how it's done in JS)
2. Access the input matrix via "rotating view" (wrapper)
3. Have four slightly different implementations for each direction

Second approach will be more readable, but it has worst performant.
Third approach is neither performant nor readable.

First approach provides the best performance, as you do rotation once (O(n^2)) and then access prepared data in a very cache-friendly way.

Btw, looking at the linked Java implementation I realized that it is an old version where traversal is done in the column-first order, which is not cache-friendly. I'll post the current implementation later.

**Comment by OzoneH3** [comment]
Updated to new version and added rotation.

It has the asymmetry that's mentioned in the other ticket.
But when looking at only one direction it's only off by ~ 1.5*intensity on the outside.
The rest seems ok though. (Maybe only because I convert to int ?)

**Comment by OzoneH3** [comment]
Some tests:
Size 100x100:

All Lights:
Total 10 tests: 3,0264600000 - Average: 0,3026460000
Total 10 tests: 2,9825410000 - Average: 0,2982541000
Total 10 tests: 3,0197960000 - Average: 0,3019796000

All Obstacles:
Total 10 tests: 0,2538550000 - Average: 0,0253855000
Total 10 tests: 0,2377400000 - Average: 0,0237740000
Total 10 tests: 0,2421330000 - Average: 0,0242133000

All Empty:
Total 10 tests: 0,2363850000 - Average: 0,0236385000
Total 10 tests: 0,2388520000 - Average: 0,0238852000
Total 10 tests: 0,2130630000 - Average: 0,0213063000

**Comment by OzoneH3** [comment]
Added obstacle transparency and light source strength but there
is some weirdness when going from 90% Transparency


to a full obstacle.


With a full obstacle it lets more light through than with 90%.

My guess is, that I have to cut another beam but not sure where exactly.

**Comment by Aivean** [comment]
Some asymmetry on the outer edges of the cone is intentional, it comes from the angles that are selected for single direction:

    for (int a = -beamsN / 2; a < beamsN / 2; a++) { // slope

As you can see, `-beamsN / 2` is included, but `beamsN / 2` is excluded. This is done so that calculations for different directions can be calculated with the same logic and rotated input/output and edges won't be counted twice.

**Comment by Aivean** [comment]
@OzoneH3 , btw, thanks for the benchmarks! I just noticed them.
One question though: "Average: 0,3026460000" <- what time unit is this?

Regarding the obstacle transparency, it needs to be a bit more complicated.  You need to determine the beam intersection area with the tile (reuse the logic from `applyLight`) and adjust your intensity reduction by that number.   E.g. beam that intersects with half-transparent obstacle by 20% should have it's intensity reduced by only 10% (50% of 20%).

**Comment by ZhilkinSerg** [comment]
I will temporary close this to clean up PR queue.

**Pull request closed without merging**